File: android/test-app-java/src/main/java/io/rover/androidtestapp/java/MainActivity.java
package io.rover.androidtestapp.java;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import io.rover.experiences.ui.containers.StandaloneExperienceHostActivity;
import io.rover.experiences.ui.navigation.ExperienceExternalNavigationEvent;

public class MainActivity extends AppCompatActivity {

    private TextView mTextMessage;

    private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = new BottomNavigationView.OnNavigationItemSelectedListener() {

        @Override
        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
            switch (item.getItemId()) {
                case R.id.navigation_home:
                    mTextMessage.setText(R.string.title_home);
                    return true;
                case R.id.navigation_dashboard:
                    mTextMessage.setText(R.string.title_dashboard);
                    return true;
                case R.id.navigation_notifications:
                    mTextMessage.setText(R.string.title_notifications);
                    return true;
            }
            return false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mTextMessage = findViewById(R.id.message);
        BottomNavigationView navigation = findViewById(R.id.navigation);
        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);

        Button testButton = findViewById(R.id.testButton);

        testButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(
                    StandaloneExperienceHostActivity.makeIntent(
                            getApplicationContext(),
                            "59e8b9d0d4459d00102c2958",
                            MyCustomExperienceActivity.class
                    )
                );
            }
        });
    }

    static public class MyCustomExperienceActivity extends StandaloneExperienceHostActivity {
        @Override
        protected void dispatchExternalNavigationEvent(ExperienceExternalNavigationEvent externalNavigationEvent) {
            if(externalNavigationEvent instanceof ExperienceExternalNavigationEvent.Custom) {
                Snackbar.make(getExperiencesView(), "Log in first mate", Snackbar.LENGTH_LONG).show();
            } else {
                super.dispatchExternalNavigationEvent(externalNavigationEvent);
            }
        }
    }
}

File: android/test-app-gcm/src/main/java/io/rover/androidtestapp/gcm/MainActivity.kt
package io.rover.androidtestapp.gcm

import android.os.Bundle
import android.provider.Settings
import android.support.design.widget.BottomNavigationView
import android.support.v7.app.AppCompatActivity
import android.widget.Button
import com.facebook.stetho.urlconnection.ByteArrayRequestEntity
import com.facebook.stetho.urlconnection.StethoURLConnectionManager
import io.rover.rover.core.data.http.AsyncTaskAndHttpUrlConnectionInterception
import io.rover.rover.core.data.http.AsyncTaskAndHttpUrlConnectionInterceptor
import io.rover.rover.core.data.http.AsyncTaskAndHttpUrlConnectionNetworkClient
import io.rover.rover.experiences.ui.containers.StandaloneExperienceHostActivity
import kotlinx.android.synthetic.main.activity_main.message
import kotlinx.android.synthetic.main.activity_main.navigation
import timber.log.Timber
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection

class MainActivity : AppCompatActivity() {

    private val mOnNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item ->
        when (item.itemId) {
            R.id.navigation_home -> {
                message.setText(R.string.title_home)
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_dashboard -> {
                message.setText(R.string.title_dashboard)
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_notifications -> {
                message.setText(R.string.title_notifications)
                return@OnNavigationItemSelectedListener true
            }
        }
        false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)

        val bluetoothName = Settings.Secure.getString(contentResolver, "bluetooth_name")
        Timber.v("BLUETOOTH NAME: ${bluetoothName}")

        val testButton: Button = findViewById(R.id.testButton)
        testButton.setOnClickListener {

            startActivity(
                StandaloneExperienceHostActivity.makeIntent(
                    applicationContext,
                    "59e8b9d0d4459d00102c2958"
                )
            )


            // demoExperienceView.viewModel = experienceViewModel

//            dataPlugin.sendEventsTask(
//                listOf(
//                    Event(
//                        hashMapOf(
//                            Pair("a key", "hi.")
//                        ),
//                        "I am event",
//                        Date(),
//                        UUID.randomUUID()
//                    )
//                ),
//                Context(
//                    null, null, null, null, null, null, null, null, null, null, null, null, null,
//                    null, null, null, null, null, null, null, null, null, null, null
//                ), null
//            ) { result ->
//                when(result) {
//                    is NetworkResult.Success -> log.e("Sent successfully!")
//                    is NetworkResult.Error -> log.e("Failed: ${result.throwable.message}")
//                }
//            }.resume()
//
//            dataPlugin.fetchStateTask { result ->
//                when(result) {
//                    is NetworkResult.Success -> log.e("DeviceState fetched successfully: ${result.response}")
//                    is NetworkResult.Error -> log.e("Failed to fetch device: ${result.throwable.message}")
//                }
//            }.resume()
        }
    }
}

/**
 * If you want to be able to see the requests made by the Rover SDK to our API in
 * [Stetho's](http://facebook.github.io/stetho/) network inspector, copy this class into your
 * application and set an instance of it on the [AsyncTaskAndHttpUrlConnectionNetworkClient] with
 * [AsyncTaskAndHttpUrlConnectionNetworkClient.registerInterceptor] (DI instructions for
 * users to follow).
 */
class StethoRoverInterceptor : AsyncTaskAndHttpUrlConnectionInterceptor {
    override fun onOpened(httpUrlConnection: HttpURLConnection, requestPath: String, body: ByteArray): AsyncTaskAndHttpUrlConnectionInterception {
        val connManager = StethoURLConnectionManager(requestPath)
        connManager.preConnect(httpUrlConnection, ByteArrayRequestEntity(body))

        return object : AsyncTaskAndHttpUrlConnectionInterception {
            override fun onConnected() {
                connManager.postConnect()
            }

            override fun onError(exception: IOException) {
                connManager.httpExchangeFailed(exception)
            }

            override fun sniffStream(source: InputStream): InputStream =
                connManager.interpretResponseStream(source)
        }
    }
}
